File: java/HebrewDate/HebrewHoliday.java
package HebrewDate;

import Date.Weekday;
import java.util.ArrayList;
import java.util.List;

public class HebrewHoliday {
	public final HebrewDate hebrewDate;
	public final int id;

	public HebrewHoliday(HebrewDate hebrewDate) {
		int year = hebrewDate.year;
		int month = hebrewDate.month;
		int day = hebrewDate.day;

		this.hebrewDate = hebrewDate;
		int computedId = HebrewHolidayId.NONE.value;

		if (month == HebrewMonth.NISAN.value) {
			if (day == 15) {
				computedId = HebrewHolidayId.PESACH.value;
			} else if (day == 21) {
				computedId = HebrewHolidayId.PESACH_VII.value;
			} else {
				HebrewDate testDate = new HebrewDate(year, HebrewMonth.NISAN.value, 27);
				if (testDate.dayOfWeek() == Weekday.FRIDAY.value) {
					if (day == 26) {
						computedId = HebrewHolidayId.YOM_HASHOAH.value;
					}
				} else if (year >= 5757 && testDate.dayOfWeek() == Weekday.SUNDAY.value) {
					if (day == 28) {
						computedId = HebrewHolidayId.YOM_HASHOAH.value;
					}
				} else if (day == 27) {
					computedId = HebrewHolidayId.YOM_HASHOAH.value;
				}
			}
		}

		if (month == HebrewMonth.IYAR.value) {
			int nominal4 = new HebrewDate(year, HebrewMonth.IYAR.value, 4).dayOfWeek();
			int zikaronDay;
			int atzmautDay;
			if (nominal4 == Weekday.THURSDAY.value) {
				zikaronDay = 3;
				atzmautDay = 4;
			} else if (nominal4 == Weekday.FRIDAY.value) {
				zikaronDay = 2;
				atzmautDay = 3;
			} else if (nominal4 == Weekday.SUNDAY.value && year >= 5764) {
				zikaronDay = 5;
				atzmautDay = 6;
			} else {
				zikaronDay = 4;
				atzmautDay = 5;
			}

			if (day == zikaronDay) {
				computedId = HebrewHolidayId.YOM_HAZIKARON.value;
			} else if (day == atzmautDay) {
				computedId = HebrewHolidayId.YOM_HAATZMAUT.value;
			}

			if (day == 14) {
				computedId = HebrewHolidayId.PESACH_SHENI.value;
			} else if (day == 18) {
				computedId = HebrewHolidayId.LAG_BOMER.value;
			} else if (day == 28) {
				computedId = HebrewHolidayId.YOM_YERUSHALAYIM.value;
			}
		}

		if (month == HebrewMonth.SIVAN.value) {
			if (day == 6) {
				computedId = HebrewHolidayId.SHAVUOT.value;
			}
		}

		if (month == HebrewMonth.TAMMUZ.value) {
			HebrewDate fastDate = new HebrewDate(year, HebrewMonth.TAMMUZ.value, 17);
			if (fastDate.dayOfWeek() == Weekday.SATURDAY.value) {
				if (day == 18) {
					computedId = HebrewHolidayId.FAST_OF_TAMMUZ.value;
				}
			} else if (day == 17) {
				computedId = HebrewHolidayId.FAST_OF_TAMMUZ.value;
			}
		}

		if (month == HebrewMonth.AV.value) {
			HebrewDate fastDate = new HebrewDate(year, HebrewMonth.AV.value, 9);
			if (fastDate.dayOfWeek() == Weekday.SATURDAY.value) {
				if (day == 10) {
					computedId = HebrewHolidayId.FAST_OF_AV.value;
				}
			} else if (day == 9) {
				computedId = HebrewHolidayId.FAST_OF_AV.value;
			}
			if (day == 15) {
				computedId = HebrewHolidayId.TU_BAV.value;
			}
		}

		if (month == HebrewMonth.TISHREI.value) {
			if (day == 1) {
				computedId = HebrewHolidayId.ROSH_HASHANA.value;
			} else if (day == 2) {
				computedId = HebrewHolidayId.ROSH_HASHANA_II.value;
			} else {
				HebrewDate tzomDate = new HebrewDate(year, HebrewMonth.TISHREI.value, 3);
				if (tzomDate.dayOfWeek() == Weekday.SATURDAY.value) {
					if (day == 4) {
						computedId = HebrewHolidayId.TZOM_GEDALIAH.value;
					}
				} else if (day == 3) {
					computedId = HebrewHolidayId.TZOM_GEDALIAH.value;
				}

				if (day == 10) {
					computedId = HebrewHolidayId.YOM_KIPPUR.value;
				} else if (day == 15) {
					computedId = HebrewHolidayId.SUKKOT.value;
				} else if (day == 21) {
					computedId = HebrewHolidayId.HOSHANA_RABA.value;
				} else if (day == 22) {
					computedId = HebrewHolidayId.SIMCHAT_TORAH_SHEMINI_ATZERETH.value;
				}
			}
		}

		if (month == HebrewMonth.KISLEV.value || month == HebrewMonth.TEVET.value) {
			if (month == HebrewMonth.KISLEV.value && day >= 25) {
				computedId = HebrewHolidayId.CHANUKKA_I.value + (day - 25);
			}

			if (month == HebrewMonth.TEVET.value) {
				if (day == 10) {
					computedId = HebrewHolidayId.FAST_OF_TEVET.value;
				}
				int kislevDays = HebrewDate.numDaysInMonth(year, HebrewMonth.KISLEV.value);
				if (kislevDays == 29 && day <= 3) {
					computedId = HebrewHolidayId.CHANUKKA_I.value + (4 + day);
				}
				if (kislevDays == 30 && day <= 2) {
					computedId = HebrewHolidayId.CHANUKKA_I.value + (5 + day);
				}
			}
		}

		if (month == HebrewMonth.SHEVAT.value && day == 15) {
			computedId = HebrewHolidayId.TU_BSHEVAT.value;
		}

		int monthEsther = HebrewDate.numMonthInYear(year);
		if (month == monthEsther) {
			HebrewDate fastDate = new HebrewDate(year, monthEsther, 13);
			if (fastDate.dayOfWeek() == Weekday.SATURDAY.value) {
				if (day == 11) {
					computedId = HebrewHolidayId.FAST_OF_ESTHER.value;
				}
			} else if (day == 13) {
				computedId = HebrewHolidayId.FAST_OF_ESTHER.value;
			}
			if (day == 14) {
				computedId = HebrewHolidayId.PURIM.value;
			} else if (day == 15) {
				computedId = HebrewHolidayId.SHUSHAN_PURIM.value;
			}
		}

		this.id = computedId;
	}

	public boolean isShabbat() {
		return hebrewDate.dayOfWeek() == 7;
	}

	public boolean isRestDay() {
		if (hebrewDate.dayOfWeek() == 7) return true;
		return id == HebrewHolidayId.ROSH_HASHANA.value
				|| id == HebrewHolidayId.ROSH_HASHANA_II.value
				|| id == HebrewHolidayId.YOM_KIPPUR.value
				|| id == HebrewHolidayId.SUKKOT.value
				|| id == HebrewHolidayId.PESACH.value
				|| id == HebrewHolidayId.PESACH_VII.value
				|| id == HebrewHolidayId.SHAVUOT.value
				|| id == HebrewHolidayId.SIMCHAT_TORAH_SHEMINI_ATZERETH.value;
	}

	public boolean isFestive() {
		return id == HebrewHolidayId.PURIM.value
				|| id == HebrewHolidayId.SHUSHAN_PURIM.value
				|| id == HebrewHolidayId.LAG_BOMER.value
				|| id == HebrewHolidayId.TU_BAV.value
				|| id == HebrewHolidayId.CHANUKKA_I.value
				|| id == HebrewHolidayId.CHANUKKA_II.value
				|| id == HebrewHolidayId.CHANUKKA_III.value
				|| id == HebrewHolidayId.CHANUKKA_IV.value
				|| id == HebrewHolidayId.CHANUKKA_V.value
				|| id == HebrewHolidayId.CHANUKKA_VI.value
				|| id == HebrewHolidayId.CHANUKKA_VII.value
				|| id == HebrewHolidayId.CHANUKKA_VIII.value;
	}

	public boolean isFast() {
		return id == HebrewHolidayId.FAST_OF_TAMMUZ.value
				|| id == HebrewHolidayId.FAST_OF_AV.value
				|| id == HebrewHolidayId.TZOM_GEDALIAH.value
				|| id == HebrewHolidayId.FAST_OF_TEVET.value
				|| id == HebrewHolidayId.FAST_OF_ESTHER.value
				|| id == HebrewHolidayId.YOM_KIPPUR.value;
	}

	public boolean isNational() {
		return id == HebrewHolidayId.YOM_HASHOAH.value
				|| id == HebrewHolidayId.YOM_HAZIKARON.value
				|| id == HebrewHolidayId.YOM_HAATZMAUT.value
				|| id == HebrewHolidayId.YOM_YERUSHALAYIM.value;
	}

	public HebrewHoliday moed() {
		int month = hebrewDate.month;
		int day = hebrewDate.day;
		if (month == HebrewMonth.NISAN.value && day >= 15 && day <= 21) {
			return new HebrewHoliday(new HebrewDate(hebrewDate.year, HebrewMonth.NISAN.value, 15));
		}
		if (month == HebrewMonth.TISHREI.value && day >= 15 && day <= 21) {
			return new HebrewHoliday(new HebrewDate(hebrewDate.year, HebrewMonth.TISHREI.value, 15));
		}
		return null;
	}

	public HebrewHoliday erev() {
		HebrewHoliday nextHoliday = new HebrewHoliday(hebrewDate.next());
		int nid = nextHoliday.id;
		if (nid == HebrewHolidayId.PESACH.value
				|| nid == HebrewHolidayId.SHAVUOT.value
				|| nid == HebrewHolidayId.ROSH_HASHANA.value
				|| nid == HebrewHolidayId.YOM_KIPPUR.value
				|| nid == HebrewHolidayId.SUKKOT.value
				|| nid == HebrewHolidayId.FAST_OF_AV.value
				|| nid == HebrewHolidayId.YOM_HAZIKARON.value
				|| nid == HebrewHolidayId.YOM_HAATZMAUT.value) {
			return nextHoliday;
		}
		return null;
	}

	public int diaspora() {
		int month = hebrewDate.month;
		int day = hebrewDate.day;

		if (month == HebrewMonth.NISAN.value) {
			if (day == 16) return DiasporaHolidayId.PESACH_II.value;
			if (day == 22) return HebrewSecondYomTovId.PESACH_VIII.value;
		}

		if (month == HebrewMonth.SIVAN.value && day == 7) {
			return HebrewSecondYomTovId.SHAVUOT_II.value;
		}

		if (month == HebrewMonth.TISHREI.value && day == 16) {
			return DiasporaHolidayId.SUKKOT_II.value;
		}

		return id;
	}

	public boolean isSolemn() {
		return id == HebrewHolidayId.YOM_HASHOAH.value || id == HebrewHolidayId.YOM_HAZIKARON.value;
	}

	public static List<HebrewDate> holidays(int year) {
		HebrewDate first = new HebrewDate(year, HebrewMonth.TISHREI.value, 1);
		HebrewDate last = new HebrewDate(year + 1, HebrewMonth.TISHREI.value, 1);
		List<HebrewDate> dates = new ArrayList<>();
		HebrewDate h = first;
		while (h.ordinal < last.ordinal) {
			if (new HebrewHoliday(h).id != HebrewHolidayId.NONE.value) {
				dates.add(h);
			}
			h = h.next();
		}
		return dates;
	}
}

File: HebrewHoliday.py
from enum import IntEnum

from Date import Weekday
from HebrewDate import HebrewDate, HebrewMonth


class HebrewHolidayId(IntEnum):
	NONE = 0
	PESACH = 1
	PESACH_VII = 3
	YOM_HASHOAH = 5
	YOM_HAZIKARON = 6
	YOM_HAATZMAUT = 7
	PESACH_SHENI = 8
	LAG_BOMER = 9
	YOM_YERUSHALAYIM = 10
	SHAVUOT = 11
	FAST_OF_TAMMUZ = 13
	FAST_OF_AV = 14
	TU_BAV = 15
	ROSH_HASHANA = 16
	ROSH_HASHANA_II = 17
	TZOM_GEDALIAH = 18
	YOM_KIPPUR = 19
	SUKKOT = 20
	HOSHANA_RABA = 22
	SHEMINI_ATZERET = 23
	SIMCHAT_TORAH = 24
	SIMCHAT_TORAH_SHEMINI_ATZERETH = 25
	CHANUKKA_I = 26
	CHANUKKA_II = 27
	CHANUKKA_III = 28
	CHANUKKA_IV = 29
	CHANUKKA_V = 30
	CHANUKKA_VI = 31
	CHANUKKA_VII = 32
	CHANUKKA_VIII = 33
	FAST_OF_TEVET = 34
	TU_BSHEVAT = 35
	FAST_OF_ESTHER = 36
	PURIM = 37
	SHUSHAN_PURIM = 38


class DiasporaHolidayId(IntEnum):
	PESACH_II = 2
	SUKKOT_II = 21


class HebrewSecondYomTovId(IntEnum):
	PESACH_VIII = 4
	SHAVUOT_II = 12


class HebrewHoliday:
	def __init__(self, hebrewDate):
		year = hebrewDate.year
		month = hebrewDate.month
		day = hebrewDate.day

		self.hebrewDate = hebrewDate
		self.id = HebrewHolidayId.NONE

		if month == HebrewMonth.NISAN:
			if day == 15:
				self.id = HebrewHolidayId.PESACH
			elif day == 21:
				self.id = HebrewHolidayId.PESACH_VII
			else:
				# 27 Nisan can only ever fall on Sun/Tue/Thu/Fri.
				# Friday -> advance to Thursday (26); Sunday -> postpone to Monday (28);
				# Tue/Thu -> observed on the 27th itself.
				testDate = HebrewDate(year, HebrewMonth.NISAN, 27)
				if testDate.dayofweek == Weekday.FRIDAY:
					if day == 26:
						self.id = HebrewHolidayId.YOM_HASHOAH
				elif year >= 5757 and testDate.dayofweek == Weekday.SUNDAY:
					if day == 28:
						self.id = HebrewHolidayId.YOM_HASHOAH
				elif day == 27:
					self.id = HebrewHolidayId.YOM_HASHOAH

		if month == HebrewMonth.IYAR:
			# 4 Iyar can only ever fall on Sun/Tue/Thu/Fri.
			# Tuesday-Wednesday: normative, no shift (4/5 Iyar).
			# Thursday-Friday: advance one day to avoid running into Shabbat (3/4 Iyar).
			# Friday-Saturday: advance two days (2/3 Iyar).
			# Sunday-Monday (since 5764): postpone one day to avoid Motzei-Shabbat
			#   preparations (5/6 Iyar); before the reform, observed 4/5 Iyar as normal.
			nominal4 = HebrewDate(year, HebrewMonth.IYAR, 4).dayofweek
			if nominal4 == Weekday.THURSDAY:
				zikaronDay, atzmautDay = 3, 4
			elif nominal4 == Weekday.FRIDAY:
				zikaronDay, atzmautDay = 2, 3
			elif nominal4 == Weekday.SUNDAY and year >= 5764:
				zikaronDay, atzmautDay = 5, 6
			else:
				zikaronDay, atzmautDay = 4, 5

			if day == zikaronDay:
				self.id = HebrewHolidayId.YOM_HAZIKARON
			elif day == atzmautDay:
				self.id = HebrewHolidayId.YOM_HAATZMAUT

			if day == 14:
				self.id = HebrewHolidayId.PESACH_SHENI
			elif day == 18:
				self.id = HebrewHolidayId.LAG_BOMER
			elif day == 28:
				self.id = HebrewHolidayId.YOM_YERUSHALAYIM

		if month == HebrewMonth.SIVAN:
			if day == 6:
				self.id = HebrewHolidayId.SHAVUOT

		if month == HebrewMonth.TAMMUZ:
			# 17 Tammuz never falls on Friday, but can fall on Shabbat; postpone to Sunday.
			fastDate = HebrewDate(year, HebrewMonth.TAMMUZ, 17)
			if fastDate.dayofweek == Weekday.SATURDAY:
				if day == 18:
					self.id = HebrewHolidayId.FAST_OF_TAMMUZ
			elif day == 17:
				self.id = HebrewHolidayId.FAST_OF_TAMMUZ

		if month == HebrewMonth.AV:
			# 9 Av never falls on Friday, but can fall on Shabbat; postpone to Sunday.
			fastDate = HebrewDate(year, HebrewMonth.AV, 9)
			if fastDate.dayofweek == Weekday.SATURDAY:
				if day == 10:
					self.id = HebrewHolidayId.FAST_OF_AV
			elif day == 9:
				self.id = HebrewHolidayId.FAST_OF_AV
			if day == 15:
				self.id = HebrewHolidayId.TU_BAV

		if month == HebrewMonth.TISHREI:
			if day == 1:
				self.id = HebrewHolidayId.ROSH_HASHANA
			elif day == 2:
				self.id = HebrewHolidayId.ROSH_HASHANA_II
			else:
				# 3 Tishrei never falls on Friday, but can fall on Shabbat; postpone to Sunday.
				tzomDate = HebrewDate(year, HebrewMonth.TISHREI, 3)
				if tzomDate.dayofweek == Weekday.SATURDAY:
					if day == 4:
						self.id = HebrewHolidayId.TZOM_GEDALIAH
				elif day == 3:
					self.id = HebrewHolidayId.TZOM_GEDALIAH

				if day == 10:
					self.id = HebrewHolidayId.YOM_KIPPUR
				elif day == 15:
					self.id = HebrewHolidayId.SUKKOT
				elif day == 21:
					self.id = HebrewHolidayId.HOSHANA_RABA
				elif day == 22:
					self.id = HebrewHolidayId.SIMCHAT_TORAH_SHEMINI_ATZERETH

		if month in (HebrewMonth.KISLEV, HebrewMonth.TEVET):
			if month == HebrewMonth.KISLEV and day >= 25:
				self.id = HebrewHolidayId(HebrewHolidayId.CHANUKKA_I + (day - 25))

			if month == HebrewMonth.TEVET:
				if day == 10:
					self.id = HebrewHolidayId.FAST_OF_TEVET
				kislevDays = HebrewDate.numdaysinmonth(year, HebrewMonth.KISLEV)
				if kislevDays == 29 and day <= 3:
					self.id = HebrewHolidayId(HebrewHolidayId.CHANUKKA_I + (4 + day))
				if kislevDays == 30 and day <= 2:
					self.id = HebrewHolidayId(HebrewHolidayId.CHANUKKA_I + (5 + day))

		if month == HebrewMonth.SHEVAT and day == 15:
			self.id = HebrewHolidayId.TU_BSHEVAT

		monthEsther = HebrewDate.numMonthInYear(year)
		if month == monthEsther:
			# 13 Adar never falls on Friday, but can fall on Shabbat; advance to Thursday.
			fastDate = HebrewDate(year, monthEsther, 13)
			if fastDate.dayofweek == Weekday.SATURDAY:
				if day == 11:
					self.id = HebrewHolidayId.FAST_OF_ESTHER
			elif day == 13:
				self.id = HebrewHolidayId.FAST_OF_ESTHER
			if day == 14:
				self.id = HebrewHolidayId.PURIM
			elif day == 15:
				self.id = HebrewHolidayId.SHUSHAN_PURIM

	@property
	def is_shabbat(self):
		return self.hebrewDate.dayofweek == 7

	@property
	def is_rest_day(self):
		if self.hebrewDate.dayofweek == 7:
			return True
		return self.id in (
			HebrewHolidayId.ROSH_HASHANA,
			HebrewHolidayId.ROSH_HASHANA_II,
			HebrewHolidayId.YOM_KIPPUR,
			HebrewHolidayId.SUKKOT,
			HebrewHolidayId.PESACH,
			HebrewHolidayId.PESACH_VII,
			HebrewHolidayId.SHAVUOT,
			HebrewHolidayId.SIMCHAT_TORAH_SHEMINI_ATZERETH,
		)

	@property
	def is_festive(self):
		return self.id in (
			HebrewHolidayId.PURIM,
			HebrewHolidayId.SHUSHAN_PURIM,
			HebrewHolidayId.LAG_BOMER,
			HebrewHolidayId.TU_BAV,
			HebrewHolidayId.CHANUKKA_I,
			HebrewHolidayId.CHANUKKA_II,
			HebrewHolidayId.CHANUKKA_III,
			HebrewHolidayId.CHANUKKA_IV,
			HebrewHolidayId.CHANUKKA_V,
			HebrewHolidayId.CHANUKKA_VI,
			HebrewHolidayId.CHANUKKA_VII,
			HebrewHolidayId.CHANUKKA_VIII,
		)

	@property
	def is_fast(self):
		return self.id in (
			HebrewHolidayId.FAST_OF_TAMMUZ,
			HebrewHolidayId.FAST_OF_AV,
			HebrewHolidayId.TZOM_GEDALIAH,
			HebrewHolidayId.FAST_OF_TEVET,
			HebrewHolidayId.FAST_OF_ESTHER,
			HebrewHolidayId.YOM_KIPPUR,
		)

	@property
	def is_national(self):
		return self.id in (
			HebrewHolidayId.YOM_HASHOAH,
			HebrewHolidayId.YOM_HAZIKARON,
			HebrewHolidayId.YOM_HAATZMAUT,
			HebrewHolidayId.YOM_YERUSHALAYIM,
		)

	@property
	def moed(self):
		month = self.hebrewDate.month
		day = self.hebrewDate.day
		if month == HebrewMonth.NISAN and 15 <= day <= 21:
			return HebrewHoliday(HebrewDate(self.hebrewDate.year, HebrewMonth.NISAN, 15))
		if month == HebrewMonth.TISHREI and 15 <= day <= 21:
			return HebrewHoliday(HebrewDate(self.hebrewDate.year, HebrewMonth.TISHREI, 15))
		return None

	@property
	def erev(self):
		if HebrewHoliday(self.hebrewDate.next).id in (
			HebrewHolidayId.PESACH,
			HebrewHolidayId.SHAVUOT,
			HebrewHolidayId.ROSH_HASHANA,
			HebrewHolidayId.YOM_KIPPUR,
			HebrewHolidayId.SUKKOT,
			HebrewHolidayId.FAST_OF_AV,
			HebrewHolidayId.YOM_HAZIKARON,
			HebrewHolidayId.YOM_HAATZMAUT,
		):
			return HebrewHoliday(self.hebrewDate.next)
		return None

	@property
	def diaspora(self):
		month = self.hebrewDate.month
		day = self.hebrewDate.day

		if month == HebrewMonth.NISAN:
			if day == 16:
				return DiasporaHolidayId.PESACH_II
			if day == 22:
				return HebrewSecondYomTovId.PESACH_VIII

		if month == HebrewMonth.SIVAN and day == 7:
			return HebrewSecondYomTovId.SHAVUOT_II

		if month == HebrewMonth.TISHREI and day == 16:
			return DiasporaHolidayId.SUKKOT_II

		return self.id

	@property
	def is_solemn(self):
		return self.id in (
			HebrewHolidayId.YOM_HASHOAH,
			HebrewHolidayId.YOM_HAZIKARON,
		)

	def __repr__(self):
		return f"HebrewHoliday({self.id.name})"

	@staticmethod
	def holidays(year):
		first = HebrewDate(year, HebrewMonth.TISHREI, 1)
		last = HebrewDate(year + 1, HebrewMonth.TISHREI, 1)
		dates = []
		h = first
		while h.ordinal < last.ordinal:
			if HebrewHoliday(h).id != HebrewHolidayId.NONE:
				dates.append(h)
			h = h.next
		return dates


class HolidayNames:
	NAMES_EN = [
		"",
		"Pesach",
		"Pesach II",
		"Pesach VII",
		"Pesach VIII",
		"Yom HaShoah",
		"Yom HaZikaron",
		"Yom HaAtzmaut",
		"Pesach Sheni",
		"Lag BaOmer",
		"Yom Yerushalayim",
		"Shavuot",
		"Shavuot II",
		"Fast of Tammuz",
		"Fast of Av",
		"Tu B'Av",
		"Rosh Hashanah",
		"Rosh Hashanah II",
		"Fast of Gedaliah",
		"Yom Kippur",
		"Sukkot",
		"Sukkot II",
		"Hoshana Rabbah",
		"Shemini Atzeret",
		"Simchat Torah",
		"Simchat Torah / Shemini Atzeret",
		"Chanukkah I",
		"Chanukkah II",
		"Chanukkah III",
		"Chanukkah IV",
		"Chanukkah V",
		"Chanukkah VI",
		"Chanukkah VII",
		"Chanukkah VIII",
		"Fast of Tevet",
		"Tu BiShvat",
		"Fast of Esther",
		"Purim",
		"Shushan Purim",
	]
	NAMES_HE = [
		"",
		"פסח",
		"פסח שני (גלות)",
		"פסח שביעי",
		"פסח שמיני (גלות)",
		"יום השואה",
		"יום הזיכרון",
		"יום העצמאות",
		"פסח שני",
		"ל״ג בעומר",
		"יום ירושלים",
		"שבועות",
		"שבועות שני (גלות)",
		"תענית תמוז",
		"תענית אב",
		"ט״ו באב",
		"ראש השנה",
		"ראש השנה ב׳",
		"צום גדליה",
		"יום כיפור",
		"סוכות",
		"סוכות שני (גלות)",
		"הושענא רבה",
		"שמיני עצרת",
		"שמחת תורה",
		"שמחת תורה / שמיני עצרת",
		"חנוכה א׳",
		"חנוכה ב׳",
		"חנוכה ג׳",
		"חנוכה ד׳",
		"חנוכה ה׳",
		"חנוכה ו׳",
		"חנוכה ז׳",
		"חנוכה ח׳",
		"תענית טבת",
		"ט״ו בשבט",
		"תענית אסתר",
		"פורים",
		"שושן פורים",
	]

	def __init__(self, lang, id):
		self.lang = lang
		self.id = id

	def name(self):
		if self.lang == 'en':
			return self.NAMES_EN[self.id]
		return self.NAMES_HE[self.id]


if __name__ == "__main__":
	import time as _systime
	from Date import Date

	now = _systime.localtime()
	today = Date(now.tm_year, now.tm_mon, now.tm_mday)
	hToday = HebrewDate.fromdate(today)

	year = hToday.year

	yearStart = HebrewDate(year, HebrewMonth.TISHREI, 1)
	yearEnd = HebrewDate(year + 1, HebrewMonth.TISHREI, 1)

	print(f"Holidays in Hebrew year {year}")
	print()

	hDate = yearStart
	while hDate.ordinal < yearEnd.ordinal:
		holiday = HebrewHoliday(hDate)
		diasporaId = holiday.diaspora
		if holiday.id != HebrewHolidayId.NONE or diasporaId != holiday.id:
			gDate = Date.fromordinal(hDate.ordinal)
			localName = HolidayNames('en', holiday.id).name() if holiday.id != HebrewHolidayId.NONE else ""
			line = f"{gDate.year}-{gDate.month:02d}-{gDate.day:02d}  {localName}"
			if diasporaId != holiday.id:
				line += f"{HolidayNames('en', diasporaId).name()} (diaspora)"
			print(line)
		hDate = hDate.next
